llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -500,6 +500,51 @@ static bool isSelect01(const APInt &C1I, const APInt &C2I) {
   return C1I.isOne() || C1I.isAllOnes() || C2I.isOne() || C2I.isAllOnes();
 }
 
+/// Try to simplify seletion chain with partially identical conditions, eg:
+///   %s1 = select i1 %c1, i32 23, i32 45
+///   %s2 = select i1 %c2, i32 666, i32 %s1
+///   %s3 = select i1 %c1, i32 789, i32 %s2
+/// -->
+///   %s2 = select i1 %c2, i32 666, i32 45
+///   %s3 = select i1 %c1, i32 789, i32 %s2
+static bool simplifySeqSelectWithSameCond(SelectInst &SI,
+                                          const SimplifyQuery &SQ,
+                                          InstCombinerImpl &IC) {
+  Value *CondVal = SI.getCondition();
+  auto trySimplifySeqSelect = [=, &SI, &IC](unsigned OpIndex) {
+    assert((OpIndex == 1 || OpIndex == 2) && "Unexpected operand index");
+    SelectInst *SINext = &SI;
+    Type *SelType = SINext->getType();
+    Value *ValOp = SINext->getOperand(OpIndex);
+    Value *CondNext;
+    // Don't need propagate FMF flag because we update the operand of SINext
+    // directly.
+    // It is not profitable to build a new select for SINext with multi-arms.
+    while (match(ValOp, m_Select(m_Value(CondNext), m_Value(), m_Value()))) {
+      if (CondNext == CondVal && SINext->hasOneUse()) {
```

Comment:
It should be `match(ValOp, m_OneUse(m_Select(....))` then drop the `SINext->hasOneUse()`, otherwise this won't fold the base select if it has multiple uses.

Reply:
For the following case, it can be optimized because **%sel1** is a sinle use node.
If we adjust according above change, then it will not optimize because the operand of **%sel1** （%sel0）is multi used.
```
define i8 @sequence_select_with_same_cond_multi_arms_src(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
  %sel0 = select i1 %cond0, i8 %a, i8 %b
  %sel1 = select i1 %cond1, i8 %sel0, i8 2; %sel1 used in single node
  %sel2 = select i1 %cond1, i8 %sel0, i8 3
  %sel3 = select i1 %cond0, i8 %sel1, i8 %sel2
  ret i8 %sel3
}

define i8 @sequence_select_with_same_cond_multi_arms_tgt(i1 %cond0, i1 %cond1, i8 %a, i8 %b) {
  %sel0 = select i1 %cond0, i8 %a, i8 %b
  %sel1 = select i1 %cond1, i8 %a, i8 2; %sel1 used in single node
  %sel2 = select i1 %cond1, i8 %sel0, i8 3
  %sel3 = select i1 %cond0, i8 %sel1, i8 %sel2
  ret i8 %sel3
}
```

Reply:
Makes sense, what about `V == SINext || SINext->hasOneUse()` then?

Reply:
do you mean **ValOp == SINext || SINext->hasOneUse()** ? the condition **ValOp == SINext** is protected by **ValOp = SINext->getOperand(OpIndex)**, and the **CondNext == CondVal** is also need, consider case. https://alive2.llvm.org/ce/z/TKLmMG
```
define i32 @src(i32 %a, i1 %c1, i1 %c2, i1 %c3){
  %s1 = select i1 %c1, i32 23, i32 45
  %s2 = select i1 %c2, i32 666, i32 %s1 ; this node can be optimized iff %c1 == %c3
  %s3 = select i1 %c3, i32 789, i32 %s2
  ret i32 %s3
}

define i32 @tgt(i32 %a, i1 %c1, i1 %c2, i1 %c3){
  %s1 = select i1 %c1, i32 23, i32 45
  %s2 = select i1 %c2, i32 666, i32 45
  %s3 = select i1 %c3, i32 789, i32 %s2
  ret i32 %s3
}
```